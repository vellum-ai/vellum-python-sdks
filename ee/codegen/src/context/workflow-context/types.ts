export interface SDK_MODULE_PATHS {
  WORKFLOWS_MODULE_PATH: readonly string[];
  CORE_NODES_MODULE_PATH: readonly string[];
  DISPLAYABLE_NODES_MODULE_PATH: readonly string[];
  INPUTS_MODULE_PATH: readonly string[];
  SANDBOX_RUNNER_MODULE_PATH: readonly string[];
  STATE_MODULE_PATH: readonly string[];
  NODE_DISPLAY_MODULE_PATH: readonly string[];
  NODE_DISPLAY_TYPES_MODULE_PATH: readonly string[];
  VELLUM_TYPES_MODULE_PATH: readonly string[];
  PORTS_MODULE_PATH: readonly string[];
}
