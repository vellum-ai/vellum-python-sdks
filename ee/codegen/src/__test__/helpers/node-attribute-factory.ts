import { NodeAttribute } from "src/types/vellum";

export const nodeAttributeFactory = (
  id: string,
  name: string,
  attributes: NodeAttribute[]
): NodeAttribute => ({
  id,
  name,
  value: {
    type: "CONSTANT_VALUE",
    value: {
      type: "JSON",
      value: attributes,
    },
  },
});
