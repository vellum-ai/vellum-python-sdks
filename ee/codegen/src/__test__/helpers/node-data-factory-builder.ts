import {
  AdornmentNode,
  CodeResourceDefinition,
  NodeAttribute,
  NodeOutput,
  NodePort,
  NodeTrigger,
  WorkflowDataNode as WorkflowDataNodeType,
} from "src/types/vellum";

export class NodeDataFactoryBuilder<T extends WorkflowDataNodeType> {
  public nodeData: T;

  constructor(nodeData: T) {
    this.nodeData = nodeData;
  }

  withPorts(ports: NodePort[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      ports: ports,
    };
    return this;
  }

  withAttributes(attributes: NodeAttribute[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      attributes,
    };
    return this;
  }

  withAdornments(adornments: AdornmentNode[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      adornments,
    };
    return this;
  }

  withOutputs(outputs: NodeOutput[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      outputs,
    };
    return this;
  }

  withBase(base: CodeResourceDefinition): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      base,
    };
    return this;
  }

  withTrigger(trigger: NodeTrigger): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      trigger,
    };
    return this;
  }

  build(): T {
    return this.nodeData;
  }
}
