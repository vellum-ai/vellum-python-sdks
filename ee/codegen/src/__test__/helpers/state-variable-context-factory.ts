import { VellumVariable } from "vellum-ai/api/types";

import { WorkflowContext } from "src/context";
import { StateVariableContext } from "src/context/state-variable-context";

export function stateVariableContextFactory({
  stateVariableData,
  workflowContext,
}: {
  stateVariableData: VellumVariable;
  workflowContext: WorkflowContext;
}): StateVariableContext {
  return new StateVariableContext({
    stateVariableData,
    workflowContext,
  });
}
