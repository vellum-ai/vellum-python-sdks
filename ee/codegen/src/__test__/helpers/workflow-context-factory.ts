import { entrypointNodeDataFactory } from "./node-data-factories";

import { WorkflowContext } from "src/context";

export function workflowContextFactory({
  absolutePathToOutputDirectory,
  moduleName,
  workflowClassName,
  workflowRawData,
  strict = true,
  disableFormatting = true,
}: Partial<WorkflowContext.Args> = {}): WorkflowContext {
  return new WorkflowContext({
    absolutePathToOutputDirectory:
      absolutePathToOutputDirectory || "./src/__tests__/",
    moduleName: moduleName || "code",
    workflowClassName: workflowClassName || "TestWorkflow",
    vellumApiKey: "<TEST_API_KEY>",
    workflowRawData: workflowRawData || {
      nodes: [entrypointNodeDataFactory()],
      edges: [],
    },
    strict,
    disableFormatting,
  });
}
