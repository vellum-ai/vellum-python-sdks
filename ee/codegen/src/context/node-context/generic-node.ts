import { BaseNodeContext } from "src/context/node-context/base";
import { PortContext } from "src/context/port-context";
import { GenericNode as GenericNodeType } from "src/types/vellum";

export class GenericNodeContext extends BaseNodeContext<GenericNodeType> {
  baseNodeClassName = "BaseNode";
  baseNodeDisplayClassName = "BaseNodeDisplay";

  getNodeOutputNamesById(): Record<string, string> {
    return Object.fromEntries(
      this.nodeData.outputs.map((output) => [output.id, output.name])
    );
  }

  createPortContexts(): PortContext[] {
    return this.nodeData.ports.map(
      (port) =>
        new PortContext({
          workflowContext: this.workflowContext,
          nodeContext: this,
          portId: port.id,
          portName: port.name,
          isDefault: port.type === "DEFAULT",
        })
    );
  }
}
