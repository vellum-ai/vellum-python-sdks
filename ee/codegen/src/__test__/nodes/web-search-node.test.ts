import { Writer } from "@fern-api/python-ast/core/Writer";
import { beforeEach, describe, expect, it } from "vitest";

import { workflowContextFactory } from "src/__test__/helpers";
import { inputVariableContextFactory } from "src/__test__/helpers/input-variable-context-factory";
import {
  nodePortFactory,
  webSearchNodeFactory,
} from "src/__test__/helpers/node-data-factories";
import { createNodeContext, WorkflowContext } from "src/context";
import { GenericNodeContext } from "src/context/node-context/generic-node";
import { GenericNode } from "src/generators/nodes/generic-node";
import { NodePort } from "src/types/vellum";

describe("WebSearchNode", () => {
  let workflowContext: WorkflowContext;
  let writer: Writer;
  let node: GenericNode;

  beforeEach(() => {
    workflowContext = workflowContextFactory({ strict: false });
    writer = new Writer();

    workflowContext.addInputVariableContext(
      inputVariableContextFactory({
        inputVariableData: {
          id: "search-query-input-id",
          key: "query",
          type: "STRING",
        },
        workflowContext,
      })
    );
  });

  describe("basic", () => {
    beforeEach(async () => {
      const nodePortData: NodePort[] = [
        nodePortFactory({
          id: "web-search-port-id",
        }),
      ];
      const nodeData = webSearchNodeFactory({ nodePorts: nodePortData });

      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as GenericNodeContext;

      node = new GenericNode({
        workflowContext,
        nodeContext,
      });
    });

    it("getNodeFile", async () => {
      node.getNodeFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it("getNodeDisplayFile", async () => {
      node.getNodeDisplayFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });

  describe("with custom attributes", () => {
    it("should generate a web search node with workflow input", async () => {
      const nodePortData: NodePort[] = [
        nodePortFactory({
          id: "web-search-port-id",
        }),
      ];

      const nodeData = webSearchNodeFactory({
        nodePorts: nodePortData,
        nodeAttributes: [
          {
            id: "custom-query-id",
            name: "query",
            value: {
              type: "WORKFLOW_INPUT",
              inputVariableId: "search-query-input-id",
            },
          },
          {
            id: "custom-api-key-id",
            name: "api_key",
            value: {
              type: "ENVIRONMENT_VARIABLE",
              environmentVariable: "CUSTOM_SERP_KEY",
            },
          },
          {
            id: "custom-num-results-id",
            name: "num_results",
            value: {
              type: "CONSTANT_VALUE",
              value: { type: "NUMBER", value: 5 },
            },
          },
          {
            id: "custom-location-id",
            name: "location",
            value: {
              type: "CONSTANT_VALUE",
              value: { type: "STRING", value: "California" },
            },
          },
        ],
      });

      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as GenericNodeContext;

      const testNode = new GenericNode({
        workflowContext,
        nodeContext,
      });

      testNode.getNodeFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });

  describe("value type variations", () => {
    it("should handle NODE_OUTPUT for query parameter", async () => {
      // Add a mock node output for the test
      workflowContext.addInputVariableContext(
        inputVariableContextFactory({
          inputVariableData: {
            id: "previous-node-output-id",
            key: "search_results",
            type: "STRING",
          },
          workflowContext,
        })
      );

      const nodeData = webSearchNodeFactory({
        nodeAttributes: [
          {
            id: "query-from-node-output",
            name: "query",
            value: {
              type: "NODE_OUTPUT",
              nodeId: "previous-search-node",
              nodeOutputId: "previous-node-output-id",
            },
          },
          {
            id: "api-key-secret",
            name: "api_key",
            value: {
              type: "ENVIRONMENT_VARIABLE",
              environmentVariable: "SERP_API_KEY",
            },
          },
          {
            id: "num-results-constant",
            name: "num_results",
            value: {
              type: "CONSTANT_VALUE",
              value: { type: "NUMBER", value: 5 },
            },
          },
          {
            id: "location-constant",
            name: "location",
            value: {
              type: "CONSTANT_VALUE",
              value: { type: "STRING", value: "New York" },
            },
          },
        ],
      });

      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as GenericNodeContext;

      const testNode = new GenericNode({
        workflowContext,
        nodeContext,
      });

      testNode.getNodeFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });

  describe("parameter validation", () => {
    it("should handle string to number conversion for num_results", async () => {
      const nodeData = webSearchNodeFactory({
        nodeAttributes: [
          {
            id: "query-constant",
            name: "query",
            value: {
              type: "CONSTANT_VALUE",
              value: { type: "STRING", value: "AI news" },
            },
          },
          {
            id: "num-results-string",
            name: "num_results",
            value: {
              type: "CONSTANT_VALUE",
              value: { type: "STRING", value: "15" },
            },
          },
        ],
      });

      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as GenericNodeContext;

      const testNode = new GenericNode({
        workflowContext,
        nodeContext,
      });

      testNode.getNodeFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });

  describe("edge cases", () => {
    it("should handle special characters in query", async () => {
      const nodeData = webSearchNodeFactory({
        nodeAttributes: [
          {
            id: "query-special-chars",
            name: "query",
            value: {
              type: "CONSTANT_VALUE",
              value: {
                type: "STRING",
                value: 'AI & ML "deep learning" (2024)',
              },
            },
          },
        ],
      });

      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as GenericNodeContext;

      const testNode = new GenericNode({
        workflowContext,
        nodeContext,
      });

      testNode.getNodeFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it("should handle minimal configuration with only required query", async () => {
      const nodeData = webSearchNodeFactory({
        nodeAttributes: [
          {
            id: "query-only",
            name: "query",
            value: {
              type: "CONSTANT_VALUE",
              value: { type: "STRING", value: "minimal search" },
            },
          },
        ],
      });

      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as GenericNodeContext;

      const testNode = new GenericNode({
        workflowContext,
        nodeContext,
      });

      testNode.getNodeFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });
});
