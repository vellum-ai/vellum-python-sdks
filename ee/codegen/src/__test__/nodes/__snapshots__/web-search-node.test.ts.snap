// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WebSearchNode > basic > getNodeDisplayFile 1`] = `
"from uuid import UUID

from vellum_ee.workflows.display.editor import NodeDisplayData, NodeDisplayPosition
from vellum_ee.workflows.display.nodes import BaseNodeDisplay
from vellum_ee.workflows.display.nodes.types import PortDisplayOverrides

from ...nodes.web_search_node import WebSearchNode


class WebSearchNodeDisplay(BaseNodeDisplay[WebSearchNode]):
    label = "Web Search Node"
    node_id = UUID("web-search-node-id")
    attribute_ids_by_name = {
        "query": UUID("query-attribute-id"),
        "api_key": UUID("api-key-attribute-id"),
        "num_results": UUID("num-results-attribute-id"),
        "location": UUID("location-attribute-id"),
    }
    port_displays = {
        WebSearchNode.Ports.default_port: PortDisplayOverrides(
            id=UUID("web-search-port-id")
        )
    }
    display_data = NodeDisplayData(
        position=NodeDisplayPosition(x=0, y=0), width=None, height=None
    )
"
`;

exports[`WebSearchNode > basic > getNodeFile 1`] = `
"from vellum.workflows.nodes.displayable.web_search_node import (
    WebSearchNode as BaseWebSearchNode,
)
from vellum.workflows.references import EnvironmentVariableReference


class WebSearchNode(BaseWebSearchNode):
    query = "latest AI developments"
    api_key = EnvironmentVariableReference(name="SERP_API_KEY")
    num_results = 10
    location = "United States"
"
`;

exports[`WebSearchNode > edge cases > should handle minimal configuration with only required query 1`] = `
"from vellum.workflows.nodes.displayable.web_search_node import (
    WebSearchNode as BaseWebSearchNode,
)


class WebSearchNode(BaseWebSearchNode):
    query = "minimal search"
"
`;

exports[`WebSearchNode > edge cases > should handle special characters in query 1`] = `
"from vellum.workflows.nodes.displayable.web_search_node import (
    WebSearchNode as BaseWebSearchNode,
)


class WebSearchNode(BaseWebSearchNode):
    query = 'AI & ML "deep learning" (2024)'
"
`;

exports[`WebSearchNode > parameter validation > should handle string to number conversion for num_results 1`] = `
"from vellum.workflows.nodes.displayable.web_search_node import (
    WebSearchNode as BaseWebSearchNode,
)


class WebSearchNode(BaseWebSearchNode):
    query = "AI news"
    num_results = "15"
"
`;

exports[`WebSearchNode > value type variations > should handle NODE_OUTPUT for query parameter 1`] = `
"from vellum.workflows.nodes.displayable.web_search_node import (
    WebSearchNode as BaseWebSearchNode,
)
from vellum.workflows.references import EnvironmentVariableReference


class WebSearchNode(BaseWebSearchNode):
    query = None
    api_key = EnvironmentVariableReference(name="SERP_API_KEY")
    num_results = 5
    location = "New York"
"
`;

exports[`WebSearchNode > with custom attributes > should generate a web search node with workflow input 1`] = `
"from vellum.workflows.nodes.displayable.web_search_node import (
    WebSearchNode as BaseWebSearchNode,
)
from vellum.workflows.references import EnvironmentVariableReference

from ..inputs import Inputs


class WebSearchNode(BaseWebSearchNode):
    query = Inputs.query
    api_key = EnvironmentVariableReference(name="CUSTOM_SERP_KEY")
    num_results = 5
    location = "California"
"
`;
